morl_baselines/multi_policy/multi_policy_moqlearning/mp_mo_q_learning.py

```diff
@@ -107,48 +106,144 @@ def get_config(self) -> dict:
             "initial_epsilon": self.initial_epsilon,
             "final_epsilon": self.final_epsilon,
             "epsilon_decay_steps": self.epsilon_decay_steps,
-            "weight_step_size": self.weights_step_size,
-            "num_weights": len(self.weights),
             "scalarization": self.scalarization.__name__,
+            "use_gpi": self.use_gpi,
+            "weight_selection_algo": self.weight_selection_algo,
+            "epsilon_ols": self.epsilon_ols,
+            "reuse_q_table": self.reuse_q_table,
+            "dyna": self.dyna,
+            "dyna_updates": self.dyna_updates,
         }
 
-    def __generate_weights(self, step_size):
-        return np.linspace((0.0, 1.0), (1.0, 0.0), int(1 / step_size) + 1, dtype=np.float32)
+    def update(self) -> None:
+        """This class does not implement the update method."""
+        pass
+
+    def _gpi_action(self, state: np.ndarray, w: np.ndarray) -> int:
+        """Get the action given by the GPI policy.
 
-    def eval_all_agents(self):
-        """Evaluate all agents and return the rewards and discounted rewards.
+        GPI(s, w) = argmax_a max_pi Q^pi(s, a, w) .
+
+        Args:
+            state: The state to get the action for.
+            weights: The weights to use for the scalarization.
 
         Returns:
-            a tuple of rewards and discounted rewards.
+            The action to take.
         """
-        discounted_rewards = []
-        rewards = []
-        for a in self.agents:
-            _, _, vec, discounted_vec = a.policy_eval(eval_env=self.env, weights=a.weights, writer=self.writer)
-            discounted_rewards.append(discounted_vec)
-            rewards.append(vec)
-        if self.log:
-            print(f"Evaluation of all agents: {rewards}")
-            print(f"discounted: {discounted_rewards}")
-        return rewards, discounted_rewards
-
-    def train(self):
-        """Train the algorithm."""
-        start_time = time.time()
-        training_epoch = int(self.num_timesteps / self.eval_freq)
-        for e in range(training_epoch):
-            print(f"Training epoch #{e}")
-            for a in self.agents:
-                a.train(
-                    start_time,
-                    total_timesteps=self.eval_freq,
-                    reset_num_timesteps=False,
+        q_vals = np.stack([policy.scalarized_q_values(state, w) for policy in self.policies])
+        _, action = np.unravel_index(np.argmax(q_vals), q_vals.shape)
+        return int(action)
+
+    @override
+    def eval(self, obs: np.array, w: Optional[np.ndarray] = None, policy_ind: Optional[int] = None) -> int:
```

Comment:
This override is dangerous because the function is made to called from outside (like in MO-gym policy evals if I remember well). And the policy_ind will not be given